design/resource-validation.md

```diff
@@ -0,0 +1,217 @@
+
+If a resource is invalid, NKG will report the error in its status.
+
+### Summary of Validation
+
+The table below summarizes the validation methods NKG will use. Any Gateway API resource will be validated by the
+following methods in order of their appearance in the table.
+
+| Name                         | Type    | Component             | Scope                   | Feedback loop for errors                                                         | Can be bypassed?                                                                     |
+|------------------------------|---------|-----------------------|-------------------------|----------------------------------------------------------------------------------|--------------------------------------------------------------------------------------|
+| CRD validation               | OpenAPI | Kubernetes API server | Structure, field values | Kubernetes API server returns any errors a response for an API call.             | Yes, if the CRDs are modified.                                                       |
+| Webhook validation           | Go code | Gateway API webhook   | Field values            | Kubernetes API server returns any errors a response for an API call.             | Yes, if the webhook is not installed, misconfigured, or running a different version. |
+| Re-run of webhook validation | Go code | NKG control plane     | Field values            | Errors are reported as Event for the resource.                                   | No                                                                                   |
+| NGK-specific validation      | Go code | NKG control plane     | Field values            | Errors are reported in the status of a resource after its creation/modification. | No                                                                                   |
+
+Notes:
+
+* The amount and the extent of the validation should allow multiple application developers to share a single NKG (User
+  story 1).
+* We expect that most of the validation problems will be caught by CRD and webhook validation and reported quickly to
+  users as a response to a Kubernetes API call (User story 2).
+
+### Evolution
+
+NKG will support more resources:
+
+- More Gateway API resources. For those, NGK will use the four validation methods from the table in the previous
+  section.
+- Introduce NKG resources. For those, NKG will use CRD validation (the rules of which are fully controlled by us). The
+  CRD validation will include the validation to prevent invalid NGINX configuration values and malicious values. Because
+  the CRD validation can be bypassed, NKG control plane will need to run the same validation rules. In addition to that,
+  NKG control plane will run any extra validation not possible to define via CRDs.
+
+We will not introduce any NKG webhook in the cluster (it adds operational complexity for the cluster admin and is a
```

Comment:
Do you mean providing our own webhook that would need to be deployed in addition to the Gateway API webhook?

Reply:
yep